client/src/Utxorpc/Client.hs

```diff
@@ -0,0 +1,154 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Utxorpc.Client
+  ( utxorpcService,
+    fromConfig,
+    withHeaders,
+  )
+where
+
+import qualified Data.ByteString as BS
+import Network.GRPC.Client (gzip, uncompressed)
+import Network.GRPC.Client.Helpers
+  ( GrpcClient (_grpcClientHeaders),
+    GrpcClientConfig,
+    UseTlsOrNot,
+    close,
+    grpcClientConfigSimple,
+    setupGrpcClient,
+    _grpcClientConfigCompression,
+  )
+import Network.GRPC.HTTP2.ProtoLens (RPC (RPC))
+import Network.HTTP2.Client (ClientIO, HostName, PortNumber)
+import Proto.Utxorpc.Build.V1.Build
+import Proto.Utxorpc.Submit.V1.Submit
+import Proto.Utxorpc.Sync.V1.Sync
+import Proto.Utxorpc.Watch.V1.Watch
+import Utxorpc.Logged (UtxorpcClientLogger, loggedSStream, loggedUnary)
+import Utxorpc.Types
+
+utxorpcService ::
+  HostName ->
+  PortNumber ->
+  UseTlsOrNot ->
+  Bool ->
+  [(BS.ByteString, BS.ByteString)] ->
+  Maybe (UtxorpcClientLogger m a) ->
+  ClientIO UtxorpcService
+utxorpcService host port tlsEnabled doCompress hdrs logger = do
+  client <- withHeaders hdrs <$> mkClient host port tlsEnabled doCompress
+  return $ fromClient client logger
+
+fromClient :: GrpcClient -> Maybe (UtxorpcClientLogger m a) -> UtxorpcService
+fromClient client logger =
+  UtxorpcService
+    (buildServiceImpl logger client)
+    (submitServiceImpl logger client)
+    (syncServiceImpl logger client)
+    (watchServiceImpl logger client)
+    (Network.GRPC.Client.Helpers.close client)
+
+mkClient :: HostName -> PortNumber -> UseTlsOrNot -> Bool -> ClientIO GrpcClient
+mkClient host port tlsEnabled doCompress = do
+  setupGrpcClient
+    ( (grpcClientConfigSimple host port tlsEnabled)
+        { _grpcClientConfigCompression = compression
+        }
+    )
+  where
+    compression = if doCompress then gzip else uncompressed
```

Comment:
An additional thought that may be helpful: it's common for this kind of library to have two main entry points, one that uses a default value for things like the port number and one that allows you to tweak all of those. Usually, the more detailed one takes a config record, which is a bit less error-prone than having to get the parameter order right. The one that takes a record usually has a name that ends with `With`. The simple one is usually implemented by filling out a record and calling the detailed one, and the function that fills in the record with default values is usually exposed as well, so that you can override specific values and then call the `With` function. This is better than having to specify all the values even when you need to override only one or two of them.

Reply:
I was trying to do something similar to the two entry points that you have described, but I must have missed the mark!

There is now:
1. `utxorpcService` (takes a `ServiceInfo` record type defined by me, which has a `default` builder)
2. `simpleUtxorpcService (takes host name, port number, use tls or not)
3. `utxorpcServiceWith` (takes a `GrpcClientConfig` record, defined in the gRPC library)

I've done it this way because the `GrpcClientConfig` specifies a lot of extra parameters, which seems a bit excessive for many use cases, but the `simple` option is nice for those not needing headers or logging.